src/linker/Linker.Dataflow/XmlFlowAnnotationSource.cs

```diff
@@ -0,0 +1,690 @@
+		}
+
+		void MatchExportedType (ExportedType exportedType, ModuleDefinition module, Regex regex, XPathNavigator nav)
+		{
+			if (regex.Match (exportedType.FullName).Success) {
+				TypeDefinition type = exportedType.Resolve ();
+				if (type != null) {
+					ProcessType (type, nav);
+				}
+			}
+		}
+
+
+		void ProcessTypePattern (string fullname, AssemblyDefinition assembly, XPathNavigator nav)
+		{
+			Regex regex = CreateRegexFromPattern (fullname);
+
+			foreach (TypeDefinition type in assembly.MainModule.Types) {
+				MatchType (type, regex, nav);
+			}
+
+			if (assembly.MainModule.HasExportedTypes) {
+				foreach (var exported in assembly.MainModule.ExportedTypes) {
+					MatchExportedType (exported, assembly.MainModule, regex, nav);
+				}
+			}
+		}
+
+		protected virtual void ProcessType (TypeDefinition type, XPathNavigator nav)
+		{
+#if !FEATURE_ILLINK
+			if (IsExcluded (nav))
+				return;
+#endif
+			ProcessTypeChildren (type, nav);
+
+			if (!type.HasNestedTypes)
+				return;
+
+			foreach (TypeDefinition nested in type.NestedTypes) {
+				var iterator = nav.SelectChildren ("type", _ns);
+				while (iterator.MoveNext ()) {
+					if (nested.Name == GetAttribute (iterator.Current, "name"))
+						ProcessTypeChildren (nested, iterator.Current);
+				}
+			}
+		}
+
+		void ProcessTypeChildren (TypeDefinition type, XPathNavigator nav)
+		{
+			if (nav.HasChildren) {
+				ProcessSelectedFields (nav, type);
+				ProcessSelectedMethods (nav, type);
+				ProcessSelectedProperties (nav, type);
+			}
+		}
+
+		void ProcessSelectedFields (XPathNavigator nav, TypeDefinition type)
+		{
+			XPathNodeIterator fields = nav.SelectChildren ("field", _ns);
+			if (fields.Count == 0)
+				return;
+
+			ProcessFields (type, fields);
+		}
+
+		void ProcessSelectedMethods (XPathNavigator nav, TypeDefinition type)
+		{
+			XPathNodeIterator methods = nav.SelectChildren ("method", _ns);
+			if (methods.Count == 0)
+				return;
+
+			ProcessMethods (type, methods);
+		}
+
+		void ProcessSelectedProperties (XPathNavigator nav, TypeDefinition type)
+		{
+			XPathNodeIterator properties = nav.SelectChildren ("property", _ns);
+			if (properties.Count == 0)
+				return;
+
+			ProcessProperties (type, properties);
+		}
+
+		static TypePreserve GetTypePreserve (XPathNavigator nav)
+		{
+			string attribute = GetAttribute (nav, _preserve);
+			if (string.IsNullOrEmpty (attribute))
+				return nav.HasChildren ? TypePreserve.Nothing : TypePreserve.All;
+
+			if (Enum.TryParse (attribute, true, out TypePreserve result))
+				return result;
+			return TypePreserve.Nothing;
+		}
+
+		void ProcessFields (TypeDefinition type, XPathNodeIterator iterator)
+		{
+			while (iterator.MoveNext ())
+				ProcessField (type, iterator);
+		}
+
+		protected virtual void ProcessField (TypeDefinition type, XPathNodeIterator iterator)
+		{
+#if !FEATURE_ILLINK
+			if (IsExcluded (iterator.Current))
+				return;
+#endif
+			string value = GetSignature (iterator.Current);
```

Comment:
There's no test coverage for this. While yes, technically one could have fields with the same name that only differ by their type, I'm not sure if we really need to have syntax to express this. Mainstream .NET languages don't support this.

Reply:
Let's not support this....